Scripts/Runtime/Data/VirtualData/VirtualData.cs

```diff
@@ -0,0 +1,296 @@
+{
+    /// <summary>
+    /// This class represents wrapped collections of data and manages them for use in Jobs.
+    /// </summary>
+    /// <remarks>
+    /// In Unity's ECS, data is stored on <see cref="Entity"/>'s via <see cref="IComponentBase"/> structs.
+    /// Unity's <see cref="SystemBase"/>'s handle the dependencies on the different sorts of data that are needed
+    /// for a given update call depending on the <see cref="EntityQuery"/>s used and Jobs scheduled.
+    ///
+    /// There are often cases where the overhead of using Entities doesn't make sense. The data is not persistent
+    /// and should exist for a period of time and then cease to exist. While this can be accomplished by Adding
+    /// or Removing Components from an Entity or spawning/destroying new entities with those components, it results
+    /// in a structural change which gets resolved at a sync point and happens on the main thread.
+    ///
+    /// Instead, using <see cref="VirtualData{TKey,TInstance}"/> can alleviate these issues while working in a similar
+    /// manner. Additional benefits are:
+    /// 
+    /// - Allowing for Shared Write
+    /// - - Multiple different jobs can write to the Pending collection at the same time.
+    /// - Fast reading via iteration or lookup
+    /// - The ability for instances of the data to write a result to different result destinations.
+    /// - - This gives implicit grouping of the data while still allowing for processing the overall set of data
+    ///     as one large set.
+    /// - - Getting access to writing to the results is handled automatically.
+    /// </remarks>
+    /// <typeparam name="TKey">The type of key to use to lookup data. Usually <see cref="Entity"/></typeparam>
+    /// <typeparam name="TInstance">The type of data to store</typeparam>
+    public class VirtualData<TKey, TInstance> : AbstractAnvilBase,
+                                                IVirtualData
+        where TKey : struct, IEquatable<TKey>
+        where TInstance : struct, ILookupData<TKey>
+    {
+        /// <summary>
+        /// The number of elements of <typeparamref name="TInstance"/> that can fit into a chunk (16kb)
+        /// This is useful for deciding on batch sizes.
+        /// </summary>
+        public static readonly int MAX_ELEMENTS_PER_CHUNK = ChunkUtil.MaxElementsPerChunk<TInstance>();
+
+        internal static VirtualData<TKey, TInstance> Create(params IVirtualData[] sources)
+        {
+            VirtualData<TKey, TInstance> virtualData = new VirtualData<TKey, TInstance>();
+
+            foreach (IVirtualData source in sources)
+            {
+                virtualData.AddSource(source);
+                source.AddResultDestination(virtualData);
+            }
+
+            return virtualData;
+        }
+
+        private readonly HashSet<IVirtualData> m_Sources = new HashSet<IVirtualData>();
+        private readonly HashSet<IVirtualData> m_ResultDestinations = new HashSet<IVirtualData>();
```

Comment:
Can we be consistent about what gets initialized in constructor vs declaration?

Reply:
Yes.... I've moved to the constructor. 

I don't know why but I always do Collections as declarations and everything else in the constructor. Weird habit.